### PR TITLE
[SPARK-28330][SQL] Support ANSI SQL: result offset clause in query expression

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This is a ANSI SQL and feature id is `F861`
```
<query expression> ::=
[ <with clause> ] <query expression body>
[ <order by clause> ] [ <result offset clause> ] [ <fetch first clause> ]

<result offset clause> ::=
OFFSET <offset row count> { ROW | ROWS }
```
For example:
```
SELECT customer_name, customer_gender FROM customer_dimension 
   WHERE occupation='Dancer' AND customer_city = 'San Francisco' ORDER BY customer_name;
    customer_name     | customer_gender
----------------------+-----------------
 Amy X. Lang          | Female
 Anna H. Li           | Female
 Brian O. Weaver      | Male
 Craig O. Pavlov      | Male
 Doug Z. Goldberg     | Male
 Harold S. Jones      | Male
 Jack E. Perkins      | Male
 Joseph W. Overstreet | Male
 Kevin . Campbell     | Male
 Raja Y. Wilson       | Male
 Samantha O. Brown    | Female
 Steve H. Gauthier    | Male
 William . Nielson    | Male
 William Z. Roy       | Male
(14 rows)

SELECT customer_name, customer_gender FROM customer_dimension 
   WHERE occupation='Dancer' AND customer_city = 'San Francisco' ORDER BY customer_name OFFSET 8;
   customer_name   | customer_gender
-------------------+-----------------
 Kevin . Campbell  | Male
 Raja Y. Wilson    | Male
 Samantha O. Brown | Female
 Steve H. Gauthier | Male
 William . Nielson | Male
 William Z. Roy    | Male
(6 rows)
```
There are some mainstream database support the syntax.
**PostgreSQL:**
https://www.postgresql.org/docs/11/queries-limit.html

**Vertica:**
https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Statements/SELECT/OFFSETClause.htm?zoom_highlight=offset

**MySQL:**
https://dev.mysql.com/doc/refman/5.6/en/select.html

The description for design:
**1**. Consider `OFFSET` as the special case of `LIMIT`. For example:
`SELECT * FROM a limit 10;` similar to `SELECT * FROM a limit 10 offset 0;`
`SELECT * FROM a offset 10;` similar to `SELECT * FROM a limit -1 offset 10;`
**2**. Because the current implement of `LIMIT` has good performance. For example:
`SELECT * FROM a limit 10;` parsed to the logic plan as below:
```
GlobalLimit (limit = 10)
    |--LocalLimit (limit = 10)
```
and then the physical plan as below:
```
GlobalLimitExec (limit = 10) // Take the first 10 rows globally
    |--LocalLimitExec (limit = 10) // Take the first 10 rows locally
```
This operator reduce massive shuffle and has good performance.
Sometimes, the logic plan transformed to the physical plan as:
```
CollectLimitExec (limit = 10) // Take the first 10 rows globally
```
If the SQL contains order by, such as `SELECT * FROM a order by c limit 10;`.
This SQL will be transformed to the physical plan as below:
```
TakeOrderedAndProjectExec (limit = 10) // Take the first 10 rows after sort globally
```

Based on this situation, this PR produces the following operations. For example:
`SELECT * FROM a limit 10 offset 10;` parsed to the logic plan as below:
```
GlobalLimit (limit = 10, offset = 10)
    |--LocalLimit (limit = 10, offset = 10)
```
and then the physical plan as below:
```
GlobalLimitExec (limit = 10, offset = 10) // Skip the first 10 rows and take the next 10 rows globally
    |--LocalLimitExec (limit = 10, offset = 10) // Take the first 20(limit + offset) rows locally
```
Sometimes, the logic plan transformed to the physical plan as:
```
CollectLimitExec (limit = 10, offset = 10) // Skip the first 10 rows and take the next 10 rows globally
```
If the SQL contains order by, such as `SELECT * FROM a order by c limit 10 offset 10;`.
This SQL will be transformed to the physical plan as below:
```
TakeOrderedAndProjectExec (limit = 10, offset 10) // Skip the first 10 rows and take the next 10 rows after sort globally
```
**3**.In addition to the above, there is a special case that is only offset but no limit. For example:
`SELECT * FROM a offset 10;` parsed to the logic plan as below:
```
GlobalLimit (limit = -1, offset = 10) // -1 means invalid
    |--LocalLimit (limit = -1, offset = 10) // -1 means invalid
```
If offset is very large, will generate a lot of overhead. So I add a configuration item `spark.sql.forceUsingOffsetWithoutLimit` to force running query when user knows the offset is small enough. The default value of `spark.sql.forceUsingOffsetWithoutLimit` is false.

Note: The origin PR to support this feature is https://github.com/apache/spark/pull/25416.
Because the origin PR too old, there exists massive conflict which is hard to resolve. So I open this new PR to support this feature.

### Why are the changes needed?
new feature

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
new UT.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 1]
```diff
@@ -41,18 +42,28 @@ trait LimitExec extends UnaryExecNode {
...
-case class CollectLimitExec(limit: Int, child: SparkPlan) extends LimitExec {
+case class CollectLimitExec(limit: Int, offset: Int, child: SparkPlan) extends LimitExec {
   override def output: Seq[Attribute] = child.output
   override def outputPartitioning: Partitioning = SinglePartition
-  override def executeCollect(): Array[InternalRow] = child.executeTake(limit)
+  override def executeCollect(): Array[InternalRow] = {
+    if (limit == Limit.INVALID_LIMIT.value) {
+      child.executeCollect().drop(offset)
+    } else {
+      child.executeTake(limit + offset).drop(offset)
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: `limit + offset` exists some robustness issues, I will improve it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 2]
```diff
@@ -41,18 +42,28 @@ trait LimitExec extends UnaryExecNode {
...
     SQLShuffleWriteMetricsReporter.createShuffleWriteMetrics(sparkContext)
   private lazy val readMetrics =
     SQLShuffleReadMetricsReporter.createShuffleReadMetrics(sparkContext)
   override lazy val metrics = readMetrics ++ writeMetrics
   protected override def doExecute(): RDD[InternalRow] = {
-    val locallyLimited = child.execute().mapPartitionsInternal(_.take(limit))
+    val locallyLimited = if (limit == Limit.INVALID_LIMIT.value) {
+      child.execute()
+    } else {
+      child.execute().mapPartitionsInternal(_.take(limit + offset))
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 3]
```diff
@@ -125,40 +138,92 @@ trait BaseLimitExec extends LimitExec with CodegenSupport {
...
+    // The counter name is already obtained by the upstream operators via `limitNotReachedChecks`.
+    // Here we have to inline it to not change its name. This is fine as we won't have many limit
+    // operators in one query.
+    ctx.addMutableState(CodeGenerator.JAVA_INT, countTerm, forceInline = true, useFreshName = false)
+    ctx.addMutableState(CodeGenerator.JAVA_INT, skipTerm, forceInline = true, useFreshName = false)
+    if (limit == Limit.INVALID_LIMIT.value) {
+      s"${consume(ctx, input)}"
+    } else {
+      s"""
+         | if ($countTerm < ${limit + offset}) {
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 4]
```diff
@@ -125,40 +138,92 @@ trait BaseLimitExec extends LimitExec with CodegenSupport {
...
 
   override def outputOrdering: Seq[SortOrder] = child.outputOrdering
 
   override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  override def doExecute(): RDD[InternalRow] = {
+    if (limit == Limit.INVALID_LIMIT.value) {
+      child.execute()
+    } else {
+      child.execute().mapPartitions { iter => iter.take(limit + offset)}
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 5]
```diff
@@ -125,40 +138,92 @@ trait BaseLimitExec extends LimitExec with CodegenSupport {
...
 
   override def outputPartitioning: Partitioning = child.outputPartitioning
 
   override def outputOrdering: Seq[SortOrder] = child.outputOrdering
+
+  override def doExecute(): RDD[InternalRow] = {
+    val rdd = if (limit == Limit.INVALID_LIMIT.value) {
+      child.execute()
+    } else {
+      child.execute().mapPartitions { iter => iter.take(limit + offset)}
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 6]
```diff
@@ -201,7 +267,7 @@ case class TakeOrderedAndProjectExec(
     val ord = new LazilyGeneratedOrdering(sortOrder, child.output)
     val localTopK: RDD[InternalRow] = {
       child.execute().map(_.copy()).mapPartitions { iter =>
-        org.apache.spark.util.collection.Utils.takeOrdered(iter, limit)(ord)
+        org.apache.spark.util.collection.Utils.takeOrdered(iter, limit + offset)(ord)
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 7]
```diff
@@ -213,7 +279,8 @@ case class TakeOrderedAndProjectExec(
         writeMetrics),
       readMetrics)
     shuffled.mapPartitions { iter =>
-      val topK = org.apache.spark.util.collection.Utils.takeOrdered(iter.map(_.copy()), limit)(ord)
+      val topK = org.apache.spark.util.collection.Utils.takeOrdered(
+        iter.map(_.copy()), limit + offset)(ord).drop(offset)
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: ditto

**Comment by beliefer** [comment]
@gatorsmile @cloud-fan @HyukjinKwon @dongjoon-hyun @findepi
new PR to support offset here.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 8]
```diff
@@ -443,6 +443,7 @@ queryOrganization
       (SORT BY sort+=sortItem (',' sort+=sortItem)*)?
       windowClause?
       (LIMIT (ALL | limit=expression))?
+      (OFFSET offset=expression)?
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: is OFFSET must be used with LIMIT? Is it legal to have OFFSET along?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 1 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: I have add check in `CheckAnalysis`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 2 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: I have add check in `CheckAnalysis`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 4 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: I have add check in `CheckAnalysis`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 3 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: I have add check in `CheckAnalysis`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 5 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: I have add check in `CheckAnalysis`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 9]
```diff
@@ -180,7 +246,7 @@ case class TakeOrderedAndProjectExec(
 
   override def executeCollect(): Array[InternalRow] = {
     val ord = new LazilyGeneratedOrdering(sortOrder, child.output)
-    val data = child.execute().map(_.copy()).takeOrdered(limit)(ord)
+    val data = child.execute().map(_.copy()).takeOrdered(limit + offset)(ord).drop(offset)
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: I have add check in `CheckAnalysis`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 6 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: I have add check in `CheckAnalysis`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 7 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: I have add check in `CheckAnalysis`.

**Comment by beliefer** [comment]
cc @cloud-fan @maropu @dongjoon-hyun

**Comment by beliefer** [comment]
retest this please

**Comment by cloud-fan** [comment]
The PR description well explains what this feature is and why we need it, but can we briefly explain the design?  e.g. what operator is created for OFFSET, how that operator is resolved, and how to execute that operator physically.

**Comment by beliefer** [comment]
> The PR description well explains what this feature is and why we need it, but can we briefly explain the design? e.g. what operator is created for OFFSET, how that operator is resolved, and how to execute that operator physically.

OK. I added these descriptions.

**Comment by skoant** [comment]
@beliefer are you still working on this PR?

**Comment by beliefer** [comment]
Currently, this work is suspended.

**Comment by cloud-fan** [comment]
Probably it's fine to have the LIMIT x + y approach, which is bad for the paging use case, but works for simple queries.

**Comment by abij** [comment]
@beliefer that's unfortunate. With Spark/Databricks and SQL jdbc endpoints getting more traction, paging would be great.

My personal example: 
SQLClient (DBeaver) connect to Databricks through jdbc to view and query tables.
Viewing a table will do a 'select * from x' and download the whole thing. While a connection to Hive thriftserver does support limit + offset and can scroll pages. DBeaver is feasing this feature to adding limit 100, 200 to queries.

How this is going to be picked up in the future.

**Pull request closed without merging**